### ISSUE TITLE
RFC Workspace file retention rules

### ISSUE THREAD

**Issue by tiborsimko** [reanahub/reana-client#610]
**Problem**

When running workflows, the workspace can host a lot of files that are currently never removed. This consumes disk space unnecessarily, and may prevent launching a large number of workflow runs due to disk space shortage.

Some of the files, such as final plots, are not very disk-space-hungry and may have a strategic value of being retained "forever". Other files are more temporary in nature, such as filtered tuples, and are consuming a lot of disk space that may not be needed for any debugging or inspection purposes after the workflow run successfully finishes.

One technique how to clean the unnecessary temporary files to save the workspace disk usage is to add some extra "cleaning steps" to the workflow DAG specification. Another technique is to use available features of the particular workflow engine such as Snakemake's [temp() output decorator](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#protected-and-temporary-files). However, there are certain disadvantages with these approaches, because the temporary files would be cleaned during workflow execution and would not be available for later inspection should such a need arise later after workflow terminates.

An alternative strategy which is proposed in this RFC is to introduce a configurable workspace file retention rule system that would allow researchers to express that some files produced by the workflow may be "garbage collected" sooner than others and that the final plots and other important temporary results should be kept for longer. It is the REANA platform who would then take care of cleaning the workspace based on the declared rules automatically via periodic workspace cleaning jobs, for example every night.

**Proposal**

(1) We propose to extend `reana.yaml` specification to introduce an optional workspace file-and-directory retention rule declaration clause such as:

```yaml
workspace:
  retention_rules:
    - myfetched/**/*.root: 1 day
    - myfiltered/*.root: 7 days
    - myfits/: 30 days
    - myplots/*.eps: 365 days
    - myplots/*.png: 365 days    
```

This declaration would instruct REANA to automatically delete ROOT files in the `myfetched` directory or any of its subdirectories (using zsh shell wildcard globbing style) one day after workflow finishes; retain ROOT files in the `myfiltered` directory for 7 days; retain everything under `myfits` directory for a month; and retain EPS and PNG plots in the `myplots` directory for a year.

(2) The workspace files and directories not matching any of the above rules would follow the global REANA workspace retention policy that is configured by the cluster administrators centrally. Such a retention policy may be "forever" for the `reana.cern.ch` instance, "30 days" for another `demo.reana.io` instance, etc.

(3) REANA would apply the workspace file retention rules after workflow execution terminates, whether successfully ("finished") or unsuccessfully ("failed"). The retention rules wouldn't obviously apply to "created", "pending", "queued", or "running" workflows.

(4) REANA would apply the above rules in the order of "earliest-applicable-rules-first", so please take care when expressing seemingly-conflicting wishes such as:

```yaml
workspace:
  retention_rules:
    - '**/*.root': 30 days
    - 'myfiltered/*.root': 7 days
```

In this example, the first rule would like to retain all ROOT files for 30 days, whilst the second rule would like to delete all ROOT files in the `myfiltered` directory after 7 days. So, by the time REANA comes to execute the first rule, the files in the `myfiltered` directory would have already been deleted by the second rule that ran earlier. 

(5) REANA would clean the workspace irrespective of whether files matching the rules were produced by the workflow in question, uploaded by the researcher manually, or created _a posteriori_ via terminal debugging. The only importance is whether some retention rule matches the file object or the directory object based on the wildcard globbing match patterns expressed in the retention rule list. 

**Question to ponder**

- Is the above `reana.yaml` file-based syntax understandable enough out of the box, without much need to consult the documentation?

- Does the above approach cover all your "remove-these-huge-files-sooner-than-my-plots" data retention use cases and needs?

- Do you think of a non-file-based retention rule scenarios that could be useful for your workflow needs? For example, keep all files less than N1 bytes forever, keep all files less than N2 bytes for 30 days, and delete all files bigger than N3 bytes after one day?

- Is the lowest granularity of a day (as illustrated in YAML snippets above) enough detailed? Would you prefer to be able to specify shorter time periods, such as an hour, and make the deletions every hour instead of daily, in case you have really very large temporary data files and/or really a lot of workflows to run in parallel?

- Is it sufficient to have the same retention rules for both "finished" workflows and "failed" workflows? Would you prefer to configure different retention rules for different statuses, longer for "finished" workflows and shorter for "failed" ones?

Please let us know any comment or suggestion you may have in order to make the workspace file retention system both easy to use and useful to tame the growing disk space size needs!

**Comment by kratsg** [comment]
Any chance we can rename this as "artifacts" (or cache) to match what is roughly what GitLabCI/GitHubActions does?

**Comment by clelange** [comment]
I think what is missing here is the default and maximum retention period. It's nice to have some kind of granularity, but I would think that most users would use something like:

```yaml
- '**/*': 365 days
```

just to be on the "safe side".

> Is the above reana.yaml file-based syntax understandable enough out of the box, without much need to consult the documentation?

I would say yes, but not everyone is aware of the `**` syntax. I think that users largely care about their main workflow results, i.e. typically their ROOT files, and there should be some comments in the examples that e.g. `'**/*.root'` will retain all ROOT files in all workflow directories.

> Does the above approach cover all your "remove-these-huge-files-sooner-than-my-plots" data retention use cases and needs?

In my opinion yes.

> Do you think of a non-file-based retention rule scenarios that could be useful for your workflow needs? For example, keep all files less than N1 bytes forever, keep all files less than N2 bytes for 30 days, and delete all files bigger than N3 bytes after one day?

I don't think that's needed since files of similar size will typically originate from the same workflow step, corresponding to a subdirectory.

> Is the lowest granularity of a day (as illustrated in YAML snippets above) enough detailed? Would you prefer to be able to specify shorter time periods, such as an hour, and make the deletions every hour instead of daily, in case you have really very large temporary data files and/or really a lot of workflows to run in parallel?

Days are probably OK, but since the unit seems to be required anyway, allowing hours or minutes would not hurt.

> Is it sufficient to have the same retention rules for both "finished" workflows and "failed" workflows? Would you prefer to configure different retention rules for different statuses, longer for "finished" workflows and shorter for "failed" ones?

I think that failed workflows should be cleaned up sooner, e.g. somewhere between 7-30 days, and for them one doesn't need file-based rules. The only reason I see to keep a failed workflow is to check the logs and that should be feasible within that time period.

**Comment by ggustavi** [comment]
Thank for adding this feature!
I think that an a priori rule of deleting outputs larger than X bytes after Y days should be applied to prevent the risk of having users submitting jobs with in the "safe" mode with infinite lifetime for all their ouputs.
I also agree of cleaning up sooner  failed workflows if note done yet.

**Comment by tiborsimko** [comment]
@kratsg @clelange @ggustavi Thanks for your comments! We've also got some private comments being OK with the proposal, so starting the implementation soon.

> Any chance we can rename this as "artifacts" (or cache) to match what is roughly what GitLabCI/GitHubActions does?

@kratsg We were thinking about this originally, but there are some slight conceptual differences which may call for using another name. For GitHub/GitLab one would define artifacts to be preserved across time (and across runs regarding cache), whilst REANA preserves all the files by default, and does not require defining artifacts as a concept. In other words, each file is an artifact, as it were, regardless of whether it was uploaded by the author, produced by the running workflow, etc. We don't distinguish the origin of files, which files were produced by which steps. The workspace file retention rules would apply to all the files equally, even to workflow inputs. So the analogy with GitHub/GitLab artifacts (or cache) wouldn't be perfect, and if we retain the same nomenclature, then perhaps it could more confuse that help?

That said, GitHub speaks about "retention period" using `retention-days` as a key for defining custom retention periods, so merging the two ideas, perhaps we could use the following syntax (if the minimum granularity of one day is sufficient):

```yaml
workspace:
  retention_days:
    - myfetched/**/*.root: 1
    - myfiltered/*.root: 7
    - myfits/: 30
    - myplots/*.eps: 365
    - myplots/*.png: 365
```

(GitLab uses `expire_in` which seems harder to synergise on.)

> I think what is missing here is the default and maximum retention period.

@clelange We thought that this would be defined by cluster administrators as a global value that is the same for everybody. E.g. default retention period 182 days, maximum user-settable retention period 365 days for the whole site. And when a user would like to use more than 365 days, an error would pop up, preventing submission of such a workflow. Will this do?

(Otherwise we could also think of allowing each user to configure their desired defaults to be applied for every run, so that defaults would be per-user rather than per-site, if the above one-size-fits-all approach to all users having the same global retention period is too broad.)

> there should be some comments in the examples that e.g. '**/*.root' will retain all ROOT files in all workflow directories.

Yes, thanks, we'll come up with a nice documentation for this.

> I think that failed workflows should be cleaned up sooner, e.g. somewhere between 7-30 days

This could be another global default for everybody defined by the cluster administrators. That is, failed workflows are retained for 31 days, finished workflows for 182 days, with a user-configurable maximum of 365 days. (Just an example.)

Obviously, if a user specifies custom retention rules for some workflow, then these rules would override the global defaults regardless of whether workflow failed or succeeded.

> I think that an a priori rule of deleting outputs larger than X bytes after Y days should be applied to prevent the risk of having users submitting jobs with in the "safe" mode with infinite lifetime for all their outputs.

@ggustave Yes, thanks, this could be part of the global rules discussed above with Clemens, enhancing it in a way that only outputs larger than X bytes would be garbage-collected. (And some REANA deployment at `reana.example.org` could configure the global default being X=0 if they would like to be very aggressive.) Although if a user produces a lot of small files, then the total disk usage could be large anyway.

P.S. Just to be clear, the current policy for `reana.cern.ch` cluster is to retain everything "forever", which may seem too liberal. However, we have a notion of disk quota per user where we can introduce strict disk quota limit counting (say 500 GB per user) so that if somebody uses more than that, they would be prevented from executing new workflows and warned about being over quota and asked to delete some past workflow runs before the new workflow executions are accepted. This should help in keeping the overall disk usage reasonable on the site, all the while being "gentle" without imposing any too harsh global systematic deletions.

**Comment by tiborsimko** [comment]
The implementation is going for:

```yaml
workspace:
  retention_days:
    myfetched/**/*.root: 1
    myfiltered/*.root: 7
    myfits/: 30
    myplots/*.eps: 365
    myplots/*.png: 365
```

using dictionary style rather than list style, so that each file name pattern (such as `myplots/*.eps`) can be present in the YAML section exactly once. Helps to prevent duplicate installations.
